src/builders.ts

```diff
@@ -851,3 +889,92 @@ export async function callReadOnlyFunction(
 
   return response.json().then(responseJson => parseReadOnlyResponse(responseJson));
 }
+
+/**
+ * Sponsored transaction options
+ *
+ * @param  {StacksTransaction} transaction - the origin-signed transaction to sponsor
+ * @param  {String} sponsorPrivateKey - the sponsor's private key
+ * @param  {BigNum} fee - the transaction fee amount to sponsor
+ * @param  {BigNum} sponsorNonce - the nonce of the sponsor account
+ * @param  {AddressHashMode} sponsorAddressHashmode - the sponsor address hashmode
+ */
+export interface SponsorOptions {
+  transaction: StacksTransaction;
+  sponsorPrivateKey: string;
+  fee?: BigNum;
+  sponsorNonce?: BigNum;
+  sponsorAddressHashmode?: AddressHashMode;
+}
+
+/**
+ * Constructs and signs a sponsored transaction as the sponsor
+ *
+ * @param  {SponsorOptions} sponsorOptions - the sponsor options object
+ *
+ * Returns a signed sponsored transaction.
+ *
+ * @return {ClarityValue}
+ */
+export async function sponsorTransaction(
+  sponsorOptions: SponsorOptions
+): Promise<StacksTransaction> {
+  const defaultOptions = {
+    fee: new BigNum(0),
+    sponsorNonce: new BigNum(0),
+    sponsorAddressHashmode: AddressHashMode.SerializeP2PKH,
+  };
+
+  const options = Object.assign(defaultOptions, sponsorOptions);
+  const network =
+    options.transaction.version === TransactionVersion.Mainnet
+      ? new StacksMainnet()
+      : new StacksTestnet();
+  const sponsorPubKey = pubKeyfromPrivKey(options.sponsorPrivateKey);
+
+  if (!sponsorOptions.fee) {
+    let txFee = new BigNum(0);
+    switch (options.transaction.payload.payloadType) {
```

Comment:
This is missing a `default` case descriptive error message, e.g. `"tx type ${n} not supported for ..."`